src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentRuntimeNodeWriter.cs

```diff
@@ -255,8 +255,15 @@ public override void WriteHtmlContent(CodeRenderingContext context, HtmlContentI
 
             // Text node
             var content = GetHtmlContent(node);
+            var renderApi = ComponentsApi.RenderTreeBuilder.AddContent;
+            if (content != null && string.IsNullOrWhiteSpace(content))
+            {
+                // We don't want newlines to be encoded in the output.
+                renderApi = ComponentsApi.RenderTreeBuilder.AddMarkupContent;
```

Comment:
ehhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh.

Reply:
does this handle cases where a newline is embedded inside some text? `<div>Hi\r\nAjay</div>` ?

Reply:
Since the thing we're trying to prevent is the visible signs of encoding the newlines, is using `AddMarkupContent` basically the only way we can prevent that?

Reply:
Curious, @rynowak why are we not able to just do `AddMarkupContent` for all HTML content in components now that we understand HTML tags and they're different IR nodes? Just curious why we have to special case whitespace is all.

Reply:
There's a perf cost to AddMarkupContent in terms of manipulating the DOM.

Reply:
> does this handle cases where a newline is embedded inside some text? <div>Hi\r\nAjay</div> ?

That doesn't go through this. That will be rewritten by `ComponentMarkupBlockPass` and will just be output as `AddMarkupContent(" <div>Hi\r\nAjay</div> ")`.

> Since the thing we're trying to prevent is the visible signs of encoding the newlines, is using AddMarkupContent basically the only way we can prevent that?

I know you were against making this `AddMarkupContent`. But there was nothing I could pass in to `AddContent` that will not be encoded. 

Just a thought, is it possible to make the implementation of `AddContent` smarter to do the right thing based on what we pass in?

Reply:
I don't think `AddContent` should be smarter. I'm wondering if we used `AddMarkupContent` for all of our static content back in the Blazor days. 

`AddContent` maps to `document.createTextNode` - ie: treat this as text that needs to be HTML encoded. https://github.com/aspnet/AspNetCore/blob/master/src/Components/Browser.JS/src/Rendering/BrowserRenderer.ts#L220

Reply:
@SteveSandersonMS do you remember how we avoided this in ye-olde-days? Do you see any way that using `AddContent` can avoid escaping the newlines?

Reply:
I just ran a blazor 0.7.0 app and it also calls `AddContent(" \n ")` . everywhere. Looks like this issue still existed in ye-olde-days.

Reply:
You need to inspect the DOM not just view source.

Reply:
Inspect Element lies. If the actual source is `<div>&#xD;&#xA; </div>` and I inspect element on it, it will still be,
```
<div>
</div>
```

Reply:
Is there another way to inspect DOM?

Reply:
@SteveSandersonMS  - here's another idea - one more layer of abstraction! If we had 3 frame kinds for text, would we be able to make the performance better?
- text that needs encoding `document.createTextNode()`
- text that does not need encoding `???`
- text that contains markup and does not need encoding `parseMarkup`

Reply:
> do you remember how we avoided this in ye-olde-days?

For text content, I think we HTML *decoded* during compilation. Then at runtime we'd call `createTextNode` with a string that contains real newlines, emojis, etc. and that does what you want. Really that's the right thing to do, because HTML encoding is only a meaningful concept for documents represented as HTML text, which for Razor Components is only true pre-compilation (or for markup blocks). However I recognize we might not want to give the compiler all the smarts needed to decode HTML entities into .NET string equivalents.

> text that needs encoding document.createTextNode()
> text that does not need encoding ???
> text that contains markup and does not need encoding parseMarkup

If it's OK, I'd like to phrase it in terms of the actual data interchange format to reduce ambiguity about which side of the exchange is doing what kinds of transformation. So on the wire, the same three cases would look like:

1. **Plain text** (use `createTextNode`)
2. **HTML-encoded text that does not contain markup** (use either [this](https://stackoverflow.com/questions/1147359/how-to-decode-html-entities-using-jquery/1395954#1395954) or [this](https://stackoverflow.com/questions/1147359/how-to-decode-html-entities-using-jquery/23596964#23596964) to convert to case 1, then use `createTextNode`)
3. **HTML-encoded text that does contain markup** (use `parseMarkup`)

We could do benchmarks, but conceptually I'd expect case 2 to be pretty much just as expensive as 3 given what the client has to do with it.

I'd still be happy with having the compiler try to get us into case 1 as often as possible (by knowing about specific HTML entities that can be pre-decoded), and then fall back on case 3 the rest of the time. For `&#xD;&#xA;`, as long as the compiler replaces this by `"\r\n"` (as in, literally characters 13 and 10, not a backslash-prefixed-`r` etc) then we are in case 1 and everything is good.

Reply:
I did my own investigation here because Steve's comments weren't making a ton of sense to me (sorry @SteveSandersonMS). The issue with entities *is the opposite case* which is why I find this so confusing.

What I didn't get about this bug earlier is that this is limited to prerendering. Everything on the client-side works as expected when use `AddContent` (no undesirable encoding), this is **just** happening on the server renderer.  

> For text content, I think we HTML decoded during compilation. Then at runtime we'd call createTextNode with a string that contains real newlines, emojis, etc. and that does what you want. 

This is what already happens when running on the client - the client sees the newline characters (13 and 10).  The reason we have this bug is that the server is much more aggressive about encoding than the client when producing pre-rendered output. This includes things like newlines and non-latin characters. It's a source of constant pain.